### PR TITLE
Local Dev Refactor (WIP)

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Local Dev Refactor

Okay, so this is a massive PR.
There is no way I would ever expect anyone to go through the whole thing and understand it.
So what I am going to do is go over the highlights of the PR. 

## Overview

The whole point of this PR is to enable developers to use their local source code to build the emulators. 
Previous iterations of this repository failed to adequately support this use case. 

This PR strives to fill the following requirements:

1. Simplify YAML configuration file
2. Allow developers to easily bind their local code into an emulated robot and modules
3. Utilize built in caching methods of pipenv, poetry, and cmake to facillitate rebuilding of source
4. Provide End to End test coverage of common OT-3 use cases

## YAML Configuration File Simplification

Previous iterations of `opentrons-emulation` required the user to specify various source fields in the yaml configuration
file. With fields like `source-type`, `source-location`, `robot-server-source-type`, `robot-server-source-location` 
it was extremely confusing what to put for each field. Plus the were scoped to each emulator, requiring a ton of repition.

To simplify this, all that has been ripped out and burned in the eternal flames of my SSD. 
`source-type` and `source-location` has been removed and replaced with 3 new fields: 

- `monorepo-source`
- `ot3-firmware-source`
- `opentrons-modules-source`

Each field can accept 3 different values. 

1. The string `latest`. This tells the system to build from the latest source code of the `main` branch from Github
2. A commit ID. This tells the system to build from the specified commit from Github.
3. A local path on your system. This tells the system to build from your local source code. 

Source type is now inferred from the value passed to the `source` fields.
Furthermore, these fields are now scoped at the top level of the yaml file and apply to all emulators that need them. 

Below is an example of the change. 

**OLD YAML CONFIG**

```yaml
robot:
  id: otie
  hardware: ot3
  source-type: remote
  source-location: latest
  robot-server-source-type: remote
  robot-server-source-location: f3146770463ad7217905ea74f37206059f61952c
  can-server-source-type: local
  can-server-source-location: /home/derek-maggio/Documents/repos/opentrons
  opentrons-hardware-source-type: remote
  opentrons-hardware-source-location: latest
  emulation-level: hardware
  exposed-port: 31950
  can-server-exposed-port: 9898
```

**NEW YAML CONFIG**

```yaml
robot:
  id: otie
  hardware: ot3
  emulation-level: hardware
  exposed-port: 31950
  can-server-exposed-port: 9898

monorepo-source: latest  # Pull latest code
ot3-firmware-source: f3146770463ad7217905ea74f37206059f61952c  # Pull this commit
opentrons-modules-source: /home/derek-maggio/Documents/repos/opentrons-modules  # Bind my local source code in
```

## Local Dev Enablement

Previously building local source code was not well supported. It was brittle, slow, and required a full rebuild of the 
Dockerfile whenever you wanted to rebuild and push your source code to the emulators. 

The Docker architecture for the containers has change drastically to fix the above issues. 
In response, the `emulation_system` Python framework has also changed to support the architecture changes. 

### Docker Architecture Changes - Builder Containers

In order to fix issues with build source code, the actual building of source code is now done inside of "builder"
containers. These builder containers live alongside your emulators, are automatically created, and do not need to be
specified anywhere by the user. The builder containers build the source code and push the executables/binaries 
into the emulator containers. 

The workflow for this is as follows:

1. The location to get source code is specified, through the various `source` fields, by the user in YAML file
2. Source code is inserted into builder containers
3. Builder containers build source code
4. Builder containers push executables/binaries into emulators

```mermaid
---
title: General Builder Container Architecture
---
flowchart LR
    local-source[Local Source]
    remote-source[Remote Source]
    source-selection{Source Selection}
    builder-container[Builder Container]
    emulator[Emulator]
    
    local-source -- User specifies folder path --- source-selection
    remote-source -- User specifies latest or commit id --- source-selection
    
    source-selection -- Insert Source --- builder-container
    builder-container -- Build source code ---builder-container
    builder-container -- Insert executables --- emulator
```

### Docker Architecture Changes - Dockerfile

To support builder containers and simplify architecture the Dockerfile has been heavily modified. 

#### Remove `local` and `remote` image distiction

Due to all building now happening inside of the builder containers, we no longer have to support `local` and `remote`
versions of the docker images. All handling for local or remote source code will be done inside of the builder containers.

#### Push build args into env vars 

Pushed `FIRMWARE_SOURCE_DOWNLOAD_LOCATION`, `OPENTRONS_SOURCE_DOWNLOAD_LOCATION`, and
`MODULE_SOURCE_DOWNLOAD_LOCATION` into containers as environment variables. This is used for e2e testing

#### Add builder container images and entrypoint files

Each builder needs it's own image and entrypoint file. 

## Add Build Caching

Since source code building is now deferred to the builder containers we can utilize any build caching. 
As long as the builder containers are not deleted their local file system will contain all build caching.

NOTE: On the initial build, no build cache will yet exist, so build times will not be shorter. 
Any subsequent builds will be.

## Provide Common OT-3 Use Case E2E Testing

To validate that this refactor works I have added End to End style testing for the common OT-3 `opentrons-emulation`
use cases. 

The use cases are as follows: 

- OT-3 Remote Source
- OT-3 Local Source
- OT-3 with Modules
 
I used the [Python Docker SDK](https://docker-py.readthedocs.io/en/stable/) to create a framework tointrospect into 
an actual emulated Docker system. From there the created containers, volumes, bind mounts, and build arguments 
are confirmed to be as expected.

The following general checks are performed:

- Builder containers are created when expected
- Any default bind mounts are always added
- If building from local source code, that bind mount from host system to build container exists
- If building from remote source code, that bind mount from host system to build container does not exist
- After building, all executables are present and correctly configured
- Build args are configured as expected

## Important things in PR

So for this PR validation of functionality will be covered by unit test and e2e tests. But there are still quite a few
areas of interest.

- [Top Level Source Code Specificication](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-115e8fcd643fcbc565180c1c0c0189ad297b4321441b44d35d7d32b14719e604R37-R45) -
These are the changes to the Pydantic specification
- [New Source Objects](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-37dde489e308c19a7db4bac60841dfb47f956ff744f357b9e029ba46f6ba1999R1) -
These new Source objects provide an abstraction layer to generate mount strings for builders and emulators. Instead of
being hardcoded inside each `service_builder` 
- [Source State Evaluation](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-37dde489e308c19a7db4bac60841dfb47f956ff744f357b9e029ba46f6ba1999R36-R73) - 
The SourceState class provides the logic to infer whether the source code should be pulled from Github or the developer's
local system
- Refactor of service builders - 
In `emulation_system/emulation_system/compose_file_creator/conversion/service_builders/` (Couldn't get a link to a folder) - 
See below for more specifics
  - Removal of healthcheck from everything except builder containers 
  (This is because all the emulator containers now don't automatically start up because the source code might not have
  been built yet)
  - Removal of `generate_command` and `generate_depends_on` methods. No longer needs 
  - Notice inside of each service_builder that the `generate_volumes` method now calls to a `Source` object's
  `generate_emulator_mount_strings` or `generate_build_mount_strings` methods.
- [Dockerfile Changes](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-f34da55ca08f1a30591d8b0b3e885bcc678537b2a9a4aadea4f190806b374ddc)
- [Refactor of Image classes](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-326b4e847af5a8a45dda7ee04273a0a4cc2351a5275abfe609bd7ca7db8feeed)
- End to End Testing Addition
  - [Github Actions Workflow](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-deaabb10671a187e9d8bc2a25d16791aa368a5a6e3d970c3d42977712b6fd60f)
  - [SystemTestDefinition Object](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-a401c19b24804899aeca87aa3804dae887f42aa7e900d1b02c0c4ced376a954f) -
  This object provides a high level description of what the expected test results are for each e2e test. It contains
  internal methods that run the actual validation logic
  - [A new way I found for defining constants](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-d8729880d447b90e3d742cb29b0eef168052cbab39eadc276e703fecc9d374a6)
- Removal of some Github Actions that aren't that useful or necessary
  - [Opentrons Emulation Github Action Tests](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-33780efe1d2a2d201d169e0f9eaa3ebc882f953f77ada469a8284f9dcd2d6d5c) - 
  this is now replace by e2e testing
  - [Test Sample Files](https://github.com/Opentrons/opentrons-emulation/pull/230/files#diff-767122a397ac244b75edcd45f7b586159b520ef3d7b5a002c524c754aeb5da74) -
  This turned out to not be very useful because the files do not change very often

### CODE REVIEW

File: emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_can_server_service_builder.py [hunk 1]
```diff
@@ -53,15 +44,10 @@ def _generate_image(self) -> str:
...
-        image_name = (
-            CANServerImages().local_firmware_image_name
-            if source_type == SourceType.LOCAL
-            else CANServerImages().remote_firmware_image_name
-        )
-        self._logging_client.log_image_name(
-            image_name, source_type, "can-server-source-type"
-        )
+        image_name = CANServerImage().image_name
+        # self._logging_client.log_image_name(
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_can_server_service_builder.py:48]: probably uncomment or remove

File: emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_input_service_builder.py [hunk 2]
```diff
@@ -85,8 +93,8 @@ def _generate_image(self) -> str:
         overhead of calculating the same thing twice.
         """
         image_name = self._container.get_image_name()
-        source_type = self._container.source_type
-        self._logging_client.log_image_name(image_name, source_type, "source-type")
+
+        # self._logging_client.log_image_name(image_name, source_type, "source-type")
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_input_service_builder.py:97]: probably uncomment or remove

File: emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_monorepo_builder_builder.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    IntermediatePorts,
+    IntermediateVolumes,
+)
+
+from ...images import MonorepoBuilderImage
+from ...utilities.shared_functions import get_build_args
+from .abstract_service_builder import AbstractServiceBuilder
+
+
+class ConcreteMonorepoBuilderBuilder(AbstractServiceBuilder):
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_monorepo_builder_builder.py:19]: this is a bad sign usually - maybe a little too many layers of indirection. Some things may be better done as functions.

File: emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_smoothie_service_builder.py [hunk 4]
```diff
@@ -54,14 +44,8 @@ def _generate_image(self) -> str:
...
-        smoothie_images = SmoothieImages()
-        source_type = self._ot2.source_type
-        image_name = (
-            smoothie_images.local_firmware_image_name
-            if source_type == SourceType.LOCAL
-            else smoothie_images.remote_firmware_image_name
-        )
-        self._logging_client.log_image_name(image_name, source_type, "source-type")
+        image_name = SmoothieImage().image_name
+        # self._logging_client.log_image_name(image_name, source_type, "source-type")
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/conversion/service_builders/concrete_smoothie_service_builder.py:48]: delete or uncomment

**Review by sfoster1 — CHANGES_REQUESTED** [review]
If this works, it works, and I'm confident it does. But we really should avoid net 10k LOC change PRs especially when it's all hand-edited, not swapping out autogenerated stuff.

I think it's worth considering also why the PR is this big. It's a big change! But could it be decomposed into smaller ones that build on each other? If not, why not? Is there a different architecture that could have been more gently transitioned?

I'm fundamentally worried that any reasonable sized change will require huge amounts of orchestration code to be altered.

I also think that it really is completely impossible to review a 10kloc PR. We need to split this up, at the very least by interactive-rebasing so it can be browsed via commit where each commit is more or less self contained and builds on each other; we could also split it into multiple PRs.

So I'm going to request changes here for essentially review purposes. We need this to be either
- multiple PRs
- a nice maintained list of commits that build on each other

Because without one of those, this really isn't reviewable.

**Comment by DerekMaggio** [comment]
Sounds good to me. I figured I would end up needing to split it up.

**Pull request closed without merging**